sqld/src/replication/replica/replicator.rs

```diff
@@ -0,0 +1,148 @@
+use std::path::PathBuf;
+use std::time::Duration;
+
+use anyhow::bail;
+use futures::StreamExt;
+use tonic::transport::Channel;
+
+use crate::replication::frame::Frame;
+use crate::replication::replica::snapshot::TempSnapshot;
+use crate::replication::FrameNo;
+use crate::rpc::replication_log::rpc::{
+    replication_log_client::ReplicationLogClient, HelloRequest, LogOffset,
+};
+use crate::rpc::replication_log::NEED_SNAPSHOT_ERROR_MSG;
+
+use super::hook::Frames;
+use super::injector::FrameInjectorHandle;
+
+const HANDSHAKE_MAX_RETRIES: usize = 100;
+
+type Client = ReplicationLogClient<Channel>;
+
+/// The `Replicator` duty is to download frames from the primary, and pass them to the injector at
+/// transaction boundaries.
+pub struct Replicator {
+    client: Client,
+    db_path: PathBuf,
+    injector: Option<FrameInjectorHandle>,
+    current_frame_no: FrameNo,
+}
+
+impl Replicator {
+    pub fn new(db_path: PathBuf, channel: Channel, uri: tonic::transport::Uri) -> Self {
+        let client = Client::with_origin(channel, uri);
+        Self {
+            client,
+            db_path,
+            injector: None,
+            current_frame_no: FrameNo::MAX,
+        }
+    }
+
+    pub async fn run(mut self) -> anyhow::Result<()> {
+        loop {
+            if self.injector.is_none() {
+                self.try_perform_handshake().await?;
+            }
+            let _ = self.replicate().await;
+            tokio::time::sleep(Duration::from_secs(1)).await;
+        }
+    }
+
+    async fn try_perform_handshake(&mut self) -> anyhow::Result<()> {
+        for _ in 0..HANDSHAKE_MAX_RETRIES {
+            tracing::info!("Attempting to perform handshake with primary.");
+            if let Ok(resp) = self.client.hello(HelloRequest {}).await {
+                let hello = resp.into_inner();
+                if let Some(applicator) = self.injector.take() {
+                    applicator.shutdown().await?;
+                }
+                let (applicator, last_applied_frame_no) =
+                    FrameInjectorHandle::new(self.db_path.clone(), hello).await?;
+                self.current_frame_no = last_applied_frame_no;
+                self.injector.replace(applicator);
+                return Ok(());
+            }
+            tokio::time::sleep(Duration::from_secs(1)).await;
+        }
+
+        bail!("couldn't connect to primary after {HANDSHAKE_MAX_RETRIES} tries ({HANDSHAKE_MAX_RETRIES} seconds)");
```

Comment:
It may take us much longer than `HANDSHAKE_MAX_RETRIES` seconds to fail, if each attempt takes nontrivial time (which is likely; network requests often take many seconds to fail).